File: PFinalQytetetJava/QytetetJava/src/modeloqytetet/Qytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
/**
 *
 * @author antonio-w10
 */
public class Qytetet {
    
    
    
    
    
    static public final int MAX_JUGADORES = 4;
    static final int NUM_SORPRESAS = 10;
    static public final int NUM_CASILLAS = 20;
    static final int PRECIO_LIBERTAD = 200;
    static final int SALDO_SALIDA = 1000;
        
    
    private ArrayList<Sorpresa> mazo; 
    Sorpresa cartaActual;
    Jugador jugadorActual;
    private ArrayList<Jugador> jugadores;
    private Tablero tablero;
    Dado dado;
    
    EstadoJuego estadoJuego;
    
    
    
    private static final Qytetet instance = new Qytetet();
    
    private Qytetet(){
        
        mazo = new ArrayList<>(); 
        
        dado = Dado.getInstance();
        
        jugadores = new ArrayList<>();
        cartaActual = null;
        jugadorActual = null;
        estadoJuego= null;
        
        
        
    }
    
    public static Qytetet getInstance() {
        return instance;
    }
    
    
    
    
    ArrayList<Sorpresa> getMazo(){
        return mazo;
    }

    public Tablero getTablero() {
        return tablero;
    }
    
    Dado getDado(){
        return dado;
    }
    
    public EstadoJuego getEstadoJuego(){
        return this.estadoJuego;
    }
    
    public int getValorDado(){
        return this.dado.getValor();
    }
    private void inicializarCartasSorpresa(){
        
        mazo.add(new Sorpresa ("Te hemos pillado con chanclas y calcetines, lo sentimos, ¡debes ir a la carcel!", tablero.getCarcel().getNumeroCasilla(), TipoSorpresa.IRACASILLA));
        mazo.add(new Sorpresa ("venga vete", 12, TipoSorpresa.IRACASILLA));
        mazo.add(new Sorpresa ("tira pa ya", 18, TipoSorpresa.IRACASILLA));
        
        mazo.add(new Sorpresa ("siiiii", 200, TipoSorpresa.PAGARCOBRAR));
        mazo.add(new Sorpresa ("nooooo", -150, TipoSorpresa.PAGARCOBRAR));
        
        mazo.add(new Sorpresa ("recibe por vivir 50", 50, TipoSorpresa.PORCASAHOTEL));
        mazo.add(new Sorpresa ("paga por vivir 50", -50, TipoSorpresa.PORCASAHOTEL));
        
        
        mazo.add(new Sorpresa ("recibe 100", 100, TipoSorpresa.PORJUGADOR));
        mazo.add(new Sorpresa ("paga 100", -100, TipoSorpresa.PORJUGADOR));
        
        mazo.add(new Sorpresa ("Sal de ahi muchacho", 0, TipoSorpresa.SALIRCARCEL));
       
        mazo.add(new Sorpresa ("Sal de ahi muchacho", 3500, TipoSorpresa.CONVERTIRME));
        mazo.add(new Sorpresa ("Sal de ahi muchacho", 4000, TipoSorpresa.CONVERTIRME));
    }
    
    private void inicializarTablero(){
        this.tablero = new Tablero();
        
    }   
    
    
    
    void actuarSiEnCasillaEdificable(){
      boolean deboPagar = this.jugadorActual.deboPagarAlquiler();
      if (deboPagar){
        this.jugadorActual.pagarAlquiler();
        
        if (this.jugadorActual.getSaldo() <= 0){
            setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
        }
      }
      
      Casilla casilla = obtenerCasillaJugadorActual();
      boolean tengoPropietario = ((Calle)casilla).tengoPropietario();
      
      if (this.estadoJuego != EstadoJuego.ALGUNJUGADORENBANCARROTA){
          if (tengoPropietario){
            setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
          }else{
            setEstadoJuego(EstadoJuego.JA_PUEDECOMPRAROGESTIONAR);
          }
      }
      
    }
    
    void actuarSiEnCasillaNoEdificable(){
       setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
       Casilla casillaActual = this.jugadorActual.getCasillaActual();
       
       if(casillaActual.getTipo() == TipoCasilla.IMPUESTO){
           this.jugadorActual.pagarImpuesto();
           
           if (this.jugadorActual.getSaldo() <= 0){
               setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
           }
           
       }else if(casillaActual.getTipo()==TipoCasilla.JUEZ){
           encarcelarJugador();
           
       }else if(casillaActual.getTipo()==TipoCasilla.SORPRESA){
           this.cartaActual = this.mazo.remove(0);
           setEstadoJuego(EstadoJuego.JA_CONSORPRESA);
           
       }
    }
    
    public void aplicarSorpresa(){
        setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        
        if (this.cartaActual.getTipo() == TipoSorpresa.SALIRCARCEL){
            this.jugadorActual.setCartaLibertad(this.cartaActual);
        }else{
            this.mazo.add(this.cartaActual);
        }
        
        
        
        if (this.cartaActual.getTipo() == TipoSorpresa.PAGARCOBRAR){
            
            
            this.jugadorActual.modificarSaldo(this.cartaActual.getValor());
            
            if (this.jugadorActual.getSaldo() <= 0){
                setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);                                
            }            
            
        }else if (this.cartaActual.getTipo() == TipoSorpresa.IRACASILLA){
            int valor = this.cartaActual.getValor();
            boolean casillaCarcel = this.tablero.esCasillaCarcel(valor);
            
            if (casillaCarcel){
                encarcelarJugador();
            }else{
                mover(valor);                
            }            
            
        }else if (this.cartaActual.getTipo() == TipoSorpresa.PORCASAHOTEL){
            int cantidad = this.cartaActual.getValor();
            int numeroTotal = this.jugadorActual.cuantasCasasHotelesTengo();
            this.jugadorActual.modificarSaldo(cantidad*numeroTotal);
            
            if (this.jugadorActual.getSaldo() <= 0){
                setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
            }            
            
        }else if (this.cartaActual.getTipo() == TipoSorpresa.PORJUGADOR){
            
            for (Jugador jugador: this.jugadores){
                if (jugador != this.jugadorActual) {
                    jugador.modificarSaldo(this.cartaActual.getValor());
                }
                
                if (jugador.saldo<=0){
                    setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
                }
                
                this.jugadorActual.modificarSaldo(-this.cartaActual.getValor());
                
                
                if (this.jugadorActual.getSaldo() <= 0){
                    setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
                }
                
                
            }
        }else if (this.cartaActual.getTipo() == TipoSorpresa.CONVERTIRME){
            Especulador especulador = this.jugadorActual.convertirme(this.cartaActual.getValor());
            
            int i = this.jugadores.indexOf(this.jugadorActual);
            
            this.jugadores.set(i, especulador);
            
            this.jugadorActual = especulador; //this.jugadorActual = this.jugadores.get(i);
            
           
            

        }
        
    }
    
    public boolean cancelarHipoteca(int numeroCasilla){
        //numCasilla corresponde a una casilla edificable 
        //que es propiedad del jugadorActual y que está hipotecada        
        
        Casilla casilla = this.tablero.obtenerCasillaNumero(numeroCasilla);
        TituloPropiedad titulo = ((Calle)casilla).getTitulo();
                        
        setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        
        boolean cancelar = this.jugadorActual.cancelarHipoteca(titulo);
        
        return cancelar;
    }
    
    public boolean comprarTituloPropiedad(){
        boolean comprado = this.jugadorActual.comprarTituloPropiedad();
        if (comprado){
            setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        }
        
        return comprado;
    }
    
    public boolean edificarCasa(int numeroCasilla){
        boolean edificada = false;
        Casilla casilla = this.tablero.obtenerCasillaNumero(numeroCasilla);
        TituloPropiedad titulo = ((Calle)casilla).getTitulo();
        edificada = this.jugadorActual.edificarCasa(titulo);
        
        if (edificada){
            setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        }
        
        return edificada;        
    }
    
    public boolean edificarHotel(int numeroCasilla){
        // 
        //LO MISMO QUE CASA
        //
        boolean edificada = false;
        Casilla casilla = this.tablero.obtenerCasillaNumero(numeroCasilla);
        TituloPropiedad titulo = ((Calle)casilla).getTitulo();
        edificada = this.jugadorActual.edificarHotel(titulo);
        
        if (edificada){
            setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        }
        
        return edificada;
    }
    
    private void encarcelarJugador(){
        if (jugadorActual.deboIrACarcel()) { //cambiado por tengocartalibertad()
            Casilla casillaCarcel = tablero.getCarcel();
            jugadorActual.irACarcel(casillaCarcel);
            setEstadoJuego(EstadoJuego.JA_ENCARCELADO);
            System.out.println("ENCARCELADO"); 
        }
        else {
            Sorpresa carta = jugadorActual.devolverCartaLibertad();
            mazo.add(carta);    
            setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
            System.out.println("ESCAPAS DE LA CARCEL"); 
        }
    }
    
    public Sorpresa getCartaActual(){
        return cartaActual;
    }
    
    public Jugador getJugadorActual(){
        return jugadorActual;
    }
    public ArrayList<Jugador> getJugadores(){
        //Jugador [2..MAX_JUGADORES]
        return jugadores;
    }
    
    
    public void hipotecarPropiedad(int numeroCasilla){
        Casilla casilla = tablero.obtenerCasillaNumero(numeroCasilla);
        TituloPropiedad titulo = ((Calle)casilla).getTitulo();
        this.jugadorActual.hipotecarPropiedad(titulo);
        setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
    }
    
    
    
    
  
    public void inicializarJuego(ArrayList<String> nombres){
        //NO se comprueba aqui nombres [2..MAX_JUGADORES]
        inicializarJugadores(nombres);
        inicializarTablero();
        inicializarCartasSorpresa();
        salidaJugadores();
    }
    private void inicializarJugadores(ArrayList<String> nombres){
        //NO se comprueba aqui nombres [2..MAX_JUGADORES]
        
        
        for (String nom: nombres){
            jugadores.add(new Jugador(nom));
        }
    }
    
    
    
    
    
    public boolean intentarSalirCarcel(MetodoSalirCarcel metodo){
        boolean libre;
        
        if (metodo == MetodoSalirCarcel.TIRANDODADO){
            int resultado = tirarDado();
            
            if(resultado >= 5){
                this.jugadorActual.setEncarcelado(false);
            }
            
        }else if (metodo == MetodoSalirCarcel.PAGANDOLIBERTAD){
            this.jugadorActual.pagarLibertad(PRECIO_LIBERTAD);
        }
        
        libre = this.jugadorActual.getEncarcelado();
        
        if (libre){
            setEstadoJuego(EstadoJuego.JA_ENCARCELADO);
        }else{
            setEstadoJuego(EstadoJuego.JA_PREPARADO);
        }
        
        return !libre;
    }
    
    public void jugar(){
        int desplazar = this.tirarDado();
        Casilla casilla = this.jugadorActual.getCasillaActual();
        Casilla casillaDestino = this.tablero.obtenerCasillaFinal(casilla, desplazar);
        
        mover(casillaDestino.getNumeroCasilla());
    }
    
    void mover(int numCasillaDestino){
        Casilla casillaInicial = this.jugadorActual.getCasillaActual();
        Casilla casillaFinal = this.tablero.obtenerCasillaNumero(numCasillaDestino);
        this.jugadorActual.setCasillaActual(casillaFinal);
        System.out.println(casillaFinal);
        if (numCasillaDestino < casillaInicial.getNumeroCasilla()){
            this.jugadorActual.modificarSaldo(SALDO_SALIDA);
        }
        
        if (casillaFinal.soyEdificable()){
            actuarSiEnCasillaEdificable();
        }else{
            actuarSiEnCasillaNoEdificable();
        }
    
    }
    
    public Casilla obtenerCasillaJugadorActual(){
        return this.jugadorActual.getCasillaActual();
    }
    
    public ArrayList<Casilla> obtenerCasillasTablero(){
        return this.tablero.getCasillas();
    }
    
    public ArrayList<Integer> obtenerPropiedadesJugador(){
        ArrayList<TituloPropiedad> propiedades = this.jugadorActual.getPropiedades();
        
        ArrayList<Integer> numeros = new ArrayList();
        
        for (int i = 0; i < NUM_CASILLAS; i++){
        
            for (TituloPropiedad prop: propiedades){
                
                if( this.tablero.getCasillas().get(i) instanceof Calle) {
                    
                    if (((Calle)this.tablero.getCasillas().get(i)).getTitulo().equals(prop)){
                        numeros.add(i);
                    }
                }            
                
            }            
        }
        
        return numeros;
    }
    
    public ArrayList<Integer> obtenerPropiedadesJugadorSegunEstadoHipoteca(boolean estadoHipoteca){
        ArrayList<TituloPropiedad> propiedades = this.jugadorActual.obtenerPropiedades(estadoHipoteca);
        ArrayList<Integer> numeros = new ArrayList();
        
        for (int i = 0; i < NUM_CASILLAS; i++){
        
            for (TituloPropiedad prop: propiedades){
                
                if( this.tablero.getCasillas().get(i) instanceof Calle) {
                    
                    if (((Calle)this.tablero.getCasillas().get(i)).getTitulo().equals(prop)){
                        numeros.add(i);
                    }
                }
            
                
            }
            
        }
        
        return numeros;
    }
    
    
    public void obtenerRanking(){
        Collections.sort(this.jugadores);
        //prueba
        for (Jugador jug: this.jugadores){
           System.out.println(jug.getNombre());
        }
    }
    
    public int obtenerSaldoJugadorActual(){
        return this.jugadorActual.getSaldo();
    }
    
    private void salidaJugadores(){
        for (Jugador jug: this.jugadores){
            jug.setCasillaActual(this.tablero.obtenerCasillaNumero(0));
        }
        
        int numero = (int) (Math.random() * this.jugadores.size());
        
        System.out.println("Jugador Salida Actual: " + numero);
        
        this.jugadorActual = this.jugadores.get(numero);
        this.estadoJuego = EstadoJuego.JA_PREPARADO;
        
    }
    
    private void setCartaActual(Sorpresa cartaActual){
        this.cartaActual = cartaActual;
    }
    
    //Preguntar cambio DCQ
    //setEstadoJuego(estadoJuego : EstadoJuego = ALGUNJUGADORENBANCARROTA) :
    public void setEstadoJuego(EstadoJuego estado){
        this.estadoJuego = estado;
    }
    
    public void siguienteJugador(){
        
        boolean encontrado = false;
        int siguiente = 0;
        int size = this.jugadores.size();
        
        for (int i = 0; i < size && !encontrado; i++ ){
            if (this.jugadorActual.equals(this.jugadores.get(i))){
                encontrado= true;
                siguiente = i + 1;
            }
        }
        
        this.jugadorActual = this.jugadores.get(siguiente % size);
        
        if (this.jugadorActual.getEncarcelado()){
            this.estadoJuego = EstadoJuego.JA_ENCARCELADOCONOPCIONDELIBERTAD;
        }else{
            this.estadoJuego = EstadoJuego.JA_PREPARADO;
        }

        
    }
    
    int tirarDado(){
        return this.dado.tirar();
    }
    
    public void venderPropiedad(int numeroCasilla){
             
        boolean vender = false;
        Casilla casilla = this.tablero.obtenerCasillaNumero(numeroCasilla);
        
        jugadorActual.venderPropiedad(casilla);
                
        
    }
  
    
    
    @Override
    public String toString() {
        return "Qytetet{" + "mazo=" + mazo + ", cartaActual=" + cartaActual + ", jugadorActual=" + jugadorActual + ", jugadores=" + jugadores + ", tablero=" + tablero + ", dado=" + dado + '}';
    }
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
}

File: PFinalQytetetRuby/QytetetRuby/lib/qytetet.rb
#encoding: UTF-8



require_relative "sorpresa"
require_relative "tablero"
require_relative "dado"
require_relative "jugador"
require_relative "estado_juego"

require "singleton"

module ModeloQytetet
  class Qytetet
    
    include Singleton
    
    attr_reader :mazo, :tablero, :dado, :jugadorActual, :jugadores
    attr_accessor :cartaActual, :estadoJuego
    
    @@MAX_JUGADORES = 4;
    @@NUM_SORPRESAS = 10;
    @@NUM_CASILLAS = 20;
    @@PRECIO_LIBERTAD = 200;
    @@SALDO_SALIDA = 1000;
    
    def initialize
      @mazo = Array.new
      
      @dado = Dado.instance
      
      @jugadores = Array.new
      @cartaActual = nil
      @jugadorActual = nil
      @estadoJuego = nil
    
    end
    
    
        
    def inicializarCartasSorpresa
      
      
      @mazo << Sorpresa.new("Coviertete", 3500, TipoSorpresa::CONVERTIRME)
      @mazo << Sorpresa.new("Covierte te", 4000, TipoSorpresa::CONVERTIRME)
      
      @mazo << Sorpresa.new("Te hemos pillado con chanclas y calcetines, lo sentimos, ¡debes ir a la carcel!", 4, TipoSorpresa::IRACASILLA)
      @mazo << Sorpresa.new("Un fan anónimo ha pagado tu fianza. Sales de la cárcel", 0, TipoSorpresa::SALIRCARCEL)
      @mazo << Sorpresa.new("venga vete", 12, TipoSorpresa::IRACASILLA)
      @mazo << Sorpresa.new("tira pa ya", 18, TipoSorpresa::IRACASILLA)
      
      @mazo << Sorpresa.new("siiiii", 200, TipoSorpresa::PAGARCOBRAR)
      @mazo << Sorpresa.new("nooooo", -150, TipoSorpresa::PAGARCOBRAR)
      
      @mazo << Sorpresa.new("recibe por vivir 50", 50, TipoSorpresa::PORCASAHOTEL)
      @mazo << Sorpresa.new("paga por vivir 50", -50, TipoSorpresa::PORCASAHOTEL)
        
        
      @mazo << Sorpresa.new("recibe 100", 100, TipoSorpresa::PORJUGADOR)
      @mazo << Sorpresa.new("paga 100", -100, TipoSorpresa::PORJUGADOR)
        
      @mazo << Sorpresa.new("Sal de ahi muchacho", 0, TipoSorpresa::SALIRCARCEL)
      
      
        
        
    end
    
    def inicializarTablero
        @tablero = Tablero.new        
    end  
    
    
    
    def actuarSiEnCasillaEdificable()
      deboPagar = @jugadorActual.deboPagarAlquiler() 
      if (deboPagar)
        @jugadorActual.pagarAlquiler()
        
        if (@jugadorActual.saldo <= 0)
            @estadoJuego=(EstadoJuego::ALGUNJUGADORENBANCARROTA)
        end
      end
      
      casilla = obtenerCasillaJugadorActual()
      tengoPropietario = casilla.tengoPropietario()
      
      if (@estadoJuego != EstadoJuego::ALGUNJUGADORENBANCARROTA)
          if (tengoPropietario)
            @estadoJuego=(EstadoJuego::JA_PUEDEGESTIONAR)
          else
            @estadoJuego=(EstadoJuego::JA_PUEDECOMPRAROGESTIONAR)
          end
      end
      
    end
    
    def actuarSiEnCasillaNoEdificable()
      @estadoJuego=(EstadoJuego::JA_PUEDEGESTIONAR)
      casillaActual = @jugadorActual.casillaActual
       
      if(casillaActual.tipoCasilla == TipoCasilla::IMPUESTO)
          @jugadorActual.pagarImpuesto()
           
      elsif(casillaActual.tipoCasilla == TipoCasilla::JUEZ)
         encarcelarJugador()
          
      elsif(casillaActual.tipoCasilla == TipoCasilla::SORPRESA)
          @cartaActual = @mazo.delete_at(0)
          print "Carta Sorpresa:  ", @cartaActual
          puts
          @estadoJuego = (EstadoJuego::JA_CONSORPRESA)
      end
    end
    
    def aplicarSorpresa()
      @estadoJuego=(EstadoJuego::JA_PUEDEGESTIONAR)
        
        if (@cartaActual.tipo == TipoSorpresa::SALIRCARCEL)
            @jugadorActual.cartaLibertad=(@cartaActual)
        else
            @mazo << (@cartaActual)
        end
        
        if (@cartaActual.tipo == TipoSorpresa::PAGARCOBRAR)            
            
            @jugadorActual.modificarSaldo(@cartaActual.valor)
            
            if (@jugadorActual.saldo <= 0)
                @estadoJuego=(EstadoJuego::ALGUNJUGADORENBANCARROTA)                               
            end            
            
        elsif (@cartaActual.tipo == TipoSorpresa::IRACASILLA)
            valor = @cartaActual.valor
            casillaCarcel = @tablero.esCasillaCarcel(valor)
            
            if (casillaCarcel)
                encarcelarJugador()
            else
                mover(valor)               
            end         
            
        elsif (@cartaActual.tipo == TipoSorpresa::PORCASAHOTEL)
            cantidad = @cartaActual.valor
            numeroTotal = @jugadorActual.cuantasCasasHotelesTengo()
            @jugadorActual.modificarSaldo(cantidad*numeroTotal)
            
            if (@jugadorActual.saldo <= 0)
                @estadoJuego=(EstadoJuego::ALGUNJUGADORENBANCARROTA)
            end          
            
        elsif (@cartaActual.tipo == TipoSorpresa::PORJUGADOR)
            
            @jugadores.each do |jugador|
                if (jugador != @jugadorActual) 
                    jugador.modificarSaldo(@cartaActual.valor)
                end
                
                if (jugador.saldo <= 0)
                    @estadoJuego=(EstadoJuego::ALGUNJUGADORENBANCARROTA)
                end
                
                @jugadorActual.modificarSaldo(-@cartaActual.valor)
                
                
                if (@jugadorActual.saldo <= 0)
                    @estadoJuego=(EstadoJuego::ALGUNJUGADORENBANCARROTA)
                end               
            end
            
        
        
        elsif (@cartaActual.tipo == TipoSorpresa::CONVERTIRME)
            especulador = @jugadorActual.convertirme(@cartaActual.valor)
            
            i = @jugadores.index(@jugadorActual)
            
            @jugadores[i]=especulador
            #@jugadores.insert(i, especulador)
            
            @jugadorActual = especulador
            
        end
        
        
        
    end
    
    def cancelarHipoteca(numeroCasilla)
      #numCasilla corresponde a una casilla edificable 
      #que es propiedad del jugadorActual y que está hipotecada        
      
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
      titulo = casilla.titulo
                        
      @estadoJuego=(EstadoJuego::JA_PUEDEGESTIONAR)
        
      cancelar = @jugadorActual.cancelarHipoteca(titulo)
        
      return cancelar
    end
    
    def comprarTituloPropiedad()
      comprado = @jugadorActual.comprarTituloPropiedad()
        if (comprado)
            @estadoJuego=(EstadoJuego::JA_PUEDEGESTIONAR)
        end
        
        return comprado
    end
    
    def edificarCasa(numeroCasilla)
      edificada = false
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
      titulo = casilla.titulo
      edificada = @jugadorActual.edificarCasa(titulo)
        
      if (edificada)
        @estadoJuego=(EstadoJuego::JA_PUEDEGESTIONAR)
      end
        
      return edificada
    end
    
    def edificarHotel(numeroCasilla)
      #LO MISMO QUE CASA
      
      edificada = false
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
      titulo = casilla.titulo
      edificada = @jugadorActual.edificarHotel(titulo)
        
      if (edificada)
          @estadoJuego=(EstadoJuego::JA_PUEDEGESTIONAR)
      end
        
      return edificada
    end
    
    def encarcelarJugador()
      if (!@jugadorActual.deboIrACarcel()) 
        casillaCarcel = @tablero.carcel
        @jugadorActual.irACarcel(casillaCarcel)
        @estadoJuego=(EstadoJuego::JA_ENCARCELADO)
        puts "ENCARCELADO"
      else
        carta = @jugadorActual.devolverCartaLibertad()
        @mazo << (carta)
        @estadoJuego=(EstadoJuego::JA_PUEDEGESTIONAR)
        puts "ESCAPAS DE LA CARCEL"
      end
    end
       
   
    def hipotecarPropiedad(numeroCasilla)
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
      titulo = casilla.titulo
      @jugadorActual.hipotecarPropiedad(titulo)
      @estadoJuego=(EstadoJuego::JA_PUEDEGESTIONAR)
    end
        
  
    def inicializarJuego(nombres)
        
        inicializarJugadores(nombres)
        inicializarTablero()
        inicializarCartasSorpresa()
        salidaJugadores()
    end
    
    def inicializarJugadores(nombres)
      
      
      nombres.each do |nom|    #tb for .. in..
        @jugadores << Jugador.new(nom)
      end
    end
    
    
    
    
    
    def intentarSalirCarcel(metodo)
              
      if (metodo == MetodoSalirCarcel::TIRANDODADO)
        resultado = tirarDado()
            
        if(resultado >= 5)
          @jugadorActual.encarcelado=(false)
        end
            
      elsif (metodo == MetodoSalirCarcel::PAGANDOLIBERTAD)
          @jugadorActual.pagarLibertad(@@PRECIO_LIBERTAD)
      end
      
      libre = @jugadorActual.encarcelado
        
      if (libre)
        @estadoJuego=(EstadoJuego::JA_ENCARCELADO)
      else
        @estadoJuego=(EstadoJuego::JA_PREPARADO)
      end
        
      return !libre
    end
    
    def jugar()
      desplazar = tirarDado()
      casilla = @jugadorActual.casillaActual
      casillaDestino = @tablero.obtenerCasillaFinal(casilla, desplazar)
        
      mover(casillaDestino.numeroCasilla)
    end
    
    def mover(numCasillaDestino)
      casillaInicial = @jugadorActual.casillaActual
      casillaFinal = @tablero.obtenerCasillaNumero(numCasillaDestino)
      @jugadorActual.casillaActual=(casillaFinal)
      
      puts casillaFinal
      
      if (numCasillaDestino < casillaInicial.numeroCasilla)
        @jugadorActual.modificarSaldo(@@SALDO_SALIDA)
      end
        
      if (casillaFinal.soyEdificable())
        actuarSiEnCasillaEdificable()
      else
        actuarSiEnCasillaNoEdificable()
      end
      
    end
    
    def obtenerCasillaJugadorActual()
      return @jugadorActual.casillaActual
    end
    
    def obtenerCasillasTablero()
        return @tablero.casillas
    end
    
    def obtenerPropiedadesJugador()
      
      propiedades = @jugadorActual.propiedades
      numeros = Array.new
      
      i = 0        
      while ( i < @@NUM_CASILLAS) 
        propiedades.each do |prop|    
          
          
          if(@tablero.casillas[i].instance_of? Calle)
            if (@tablero.casillas[i].titulo.equal?(prop))
              numeros << (i)
            end
          end
          
          
          
        end   
        i += 1
      end
        
      return numeros;
        
    end
    
    def obtenerPropiedadesJugadorSegunEstadoHipoteca(estadoHipoteca)
      
      propiedades = @jugadorActual.obtenerPropiedades(estadoHipoteca)
      numeros = Array.new        
      
      i = 0        
      while ( i < @@NUM_CASILLAS) 
        propiedades.each do |prop|                    
          if(@tablero.casillas[i].instance_of? Calle)
            if (@tablero.casillas[i].titulo.equal?(prop))
              numeros << (i)
            end
          end
        end   
        i += 1
      end
        
      return numeros;
        
    end
    
    
    #def jugadorActualEnCalleLibre()
    #  return (@jugadorActual.casillaActual.soyEdificable() && @jugadorActual.casillaActual().tengoPropietario())
    #end
    
    #def jugadorActualEncarcelado()
    #  return @jugadorActual.encarcelado 
    #end
    
    
    def obtenerRanking()
      @jugadores=@jugadores.sort
      #prueba
        @jugadores.each do |jug|
          puts (jug.nombre)
          
        end
    end
    
    def obtenerSaldoJugadorActual()
      return @jugadorActual.saldo
    end
    
    def salidaJugadores()
      
      @jugadores.each do |jug|
        jug.casillaActual = @tablero.obtenerCasillaNumero(0);
      end
      
            
      @jugadorActual = @jugadores[rand(@jugadores.count)]
      
      print "Jugador Salida Actual: ", @jugadorActual.nombre
      puts
      @estadoJuego = EstadoJuego::JA_PREPARADO
        
    end
    
    
    def siguienteJugador()
      
      encontrado = false
      siguiente = 0
      size = @jugadores.length()
      i = 0
      while (i < size && !encontrado)
        
        if (@jugadorActual.equal?(@jugadores[i]))
          encontrado= true
          siguiente = i + 1
        end
        
        i += 1
        
      end
        
      @jugadorActual = @jugadores[(siguiente % size)]
        
      if (@jugadorActual.encarcelado)
        @estadoJuego = EstadoJuego::JA_ENCARCELADOCONOPCIONDELIBERTAD
      else
        @estadoJuego = EstadoJuego::JA_PREPARADO
      end
        
      
    end
    
    def tirarDado()
      return @dado.tirar()
    end
    
    def getValorDado()
      return @dado.valor
    end
    
    def venderPropiedad(numeroCasilla)
      vender = false
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
        
      @jugadorActual.venderPropiedad(casilla)
    end

    private :inicializarJugadores, :inicializarTablero, :inicializarCartasSorpresa, :encarcelarJugador, :salidaJugadores
    
    
    
    
    def to_s
      "mazo=#{@mazo} \n cartaActual= #{@cartaActual} \n jugadorActual= #{@jugadorActual} \n jugadores= #{@jugadores} \n tablero= #{@tablero} \n dado= #{@dado}"  
    end 
    
  end
end

